lib/offline.js

```diff
@@ -55,60 +61,121 @@ module.exports = function(config) {
-        }
+    function flatGlobs(fileGlobs) {
+      return Object.keys(fileGlobs.reduce(function (matches, fileGlob) {
+        fileGlob = fileGlob.replace(path.sep, '/');
+        glob.sync(fileGlob, { nodir: true }).forEach(function (filepath) {
+          matches[filepath] = filepath;
+        });
+        return matches;
+      }, {}));
+    }
+
+    function getFilesAndHashes(files, rootDir, sizeWhiteList) {
+      var totalSize = 0;
+      var filesAndHashes = files.map(function (filepath) {
+        var size = countFileSize(filepath, sizeWhiteList);
+        totalSize += size;
+        var data = fs.readFileSync(filepath);
+        var hash = getHash(data);
+        logOk(filepath, size);
+        return {
+          path: filepath.replace(rootDir, function (match, offset) {
+            //XXX: The root must be the worker's directory
+            return offset === 0 ? './' : match;
+          }),
+          hash: hash,
+        };
       });
 
-      return swPrecache.write(path.join(rootDir, 'offline-worker.js'), {
-        staticFileGlobs: staticFileGlobs,
-        stripPrefix: rootDir,
-        verbose: true,
-        logger: function(message) {
-          if (message.indexOf('Caching') === 0) {
-            message = chalk.bold.green('✓ ') + message.replace('static resource ', '');
-          }
-
-          console.log(message);
-        },
-        importScripts: importScripts,
-        cacheId: cacheId,
-        ignoreUrlParametersMatching: config.ignoreUrlParametersMatching || [/./],
-        maximumFileSizeToCacheInBytes: Infinity,
+      console.log('Total cache size is ' + prettyBytes(totalSize) +
+                  ' for ' + files.length + ' files.\n');
+
+      return filesAndHashes;
+    }
+
+    function checkImports(imports, rootDir) {
+      imports.forEach(function (filepath) {
+        assertExists(path.join(rootDir, filepath));
       });
-    }));
+    }
+
+    function pluckHashes(entries) {
+      return entries.map(function (entry) { return entry.hash; });
+    }
+
+    function assertExists(filepath) {
+      var stat;
+      try {
+        stat = fs.statSync(filepath);
+      } catch (ex) {
+        console.log(chalk.red.bold(filepath + ' doesn\'t exist.'));
+        throw ex;
+      }
+
+      if (!stat.isFile()) {
+        console.log(chalk.red.bold(filepath + ' is not a file.'));
+        throw new Error(filepath + ' is not a file.');
+      }
+    }
+
+    function countFileSize(filepath, whiteList) {
+      var stat = fs.statSync(filepath);
+      if (stat.isFile() && stat.size > 2 * 1024 * 1024 && whiteList.indexOf(filepath) === -1) {
+        console.log(chalk.yellow.bold(filepath + ' is bigger than 2 MiB. Are you sure you want to cache it? To suppress this warning, explicitly include the file in the fileGlobs list.'));
+      }
+      return stat.size;
+    }
+
+    function logOk(filepath, size) {
+      console.log(chalk.bold.green('✓ ') + 'Caching ' + filepath + ' (' + prettyBytes(size) + ')');
```

Comment:
Was this originally being called from multiple call sites? I'm just wondering why it's factored out into a function when it's a one-liner that is only being called from one site.